src/ingest/src/main/scala/osmesa/ingest/GenerateVT.scala

```diff
@@ -0,0 +1,74 @@
+package osmesa
+
+import com.amazonaws.services.s3.model.CannedAccessControlList._
+import geotrellis.raster._
+import geotrellis.raster.rasterize._
+import geotrellis.raster.rasterize.polygon._
+import geotrellis.spark._
+import geotrellis.spark.io.s3._
+import geotrellis.spark.tiling._
+import geotrellis.vector._
+import geotrellis.vectortile._
+import org.apache.spark.rdd.RDD
+
+object GenerateVT {
+
+  type VTF[G <: Geometry] = Feature[G, Map[String, Value]]
+
+  def save(vectorTiles: RDD[(SpatialKey, VectorTile)], zoom: Int, bucket: String, prefix: String) = {
+    vectorTiles
+      .mapValues(_.toBytes)
+      .saveToS3({ sk: SpatialKey => s"s3://${bucket}/${prefix}/${zoom}/${sk.col}/${sk.row}.mvt" },
+                putObjectModifier = { o => o.withCannedAcl(PublicRead) })
+  }
+
+  def apply[G <: Geometry](features: RDD[VTF[G]], layout: LayoutDefinition, layerName: String): RDD[(SpatialKey, VectorTile)] = {
+    val keyedGeoms = features.flatMap{ feat =>
+      val g = feat.geom
+      val keys = layout.mapTransform.keysForGeometry(g)
+      keys.map{ k => (k, feat) }
+    }
+
+    type FeatureTup = (Seq[VTF[Point]], Seq[VTF[Line]], Seq[VTF[MultiLine]], Seq[VTF[Polygon]], Seq[VTF[MultiPolygon]])
+
+    def create(feat: VTF[Geometry]): FeatureTup = feat.geom match {
+      case pt: Point => (Seq(PointFeature(pt, feat.data)), Seq.empty, Seq.empty, Seq.empty, Seq.empty)
+      case l: Line => (Seq.empty, Seq(LineFeature(l, feat.data)), Seq.empty, Seq.empty, Seq.empty)
+      case ml: MultiLine => (Seq.empty, Seq.empty, Seq(MultiLineFeature(ml, feat.data)), Seq.empty, Seq.empty)
+      case p: Polygon => (Seq.empty, Seq.empty, Seq.empty, Seq(PolygonFeature(p, feat.data)), Seq.empty)
+      case mp: MultiPolygon => (Seq.empty, Seq.empty, Seq.empty, Seq.empty, Seq(MultiPolygonFeature(mp, feat.data)))
+    }
+
+    def merge(tup: FeatureTup, feat: VTF[Geometry]) =
+      combine(tup, create(feat))
+
+    def combine(tup1: FeatureTup, tup2: FeatureTup) = {
+      val (pt1, l1, ml1, p1, mp1) = tup1
+      val (pt2, l2, ml2, p2, mp2) = tup2
+      (pt1 ++ pt2, l1 ++ l2, ml1 ++ ml2, p1 ++ p2, mp1 ++ mp2)
+    }
+
+    keyedGeoms
+      .combineByKey(create, merge, combine)
+      .map { case (sk, tup) => {
+        val (pts, ls, mls, ps, mps) = tup
+        val extent = layout.mapTransform(sk)
+
+        val layer = StrictLayer(
```

Comment:
Does this clip geometries to tile bounds?

Reply:
(Sorry if this is a dup, something was weird with github, and my last comment has vanished)

Yes.  This will quickly assign spatial keys to geometries, but will not clip to the tile.

Reply:
Specifically, line 28
```scala
val keys = layout.mapTransform.keysForGeometry(g)
```
provides the keys.

Reply:
Gotcha.

Does `StrictLayer` do any clipping when producing the MVT layer? (If not, my concern is that individual tiles will be **much** heavier than they need to be, including geometry components outside the extent of a given tile.)

A lesser concern (to potentially be borne out in testing, so maybe this is moot) is that duplicating full geometries for each key (esp. for large geometries) and then grouping by key will produce large, unbalanced partitions and more shuffle than is necessary.

Reply:
It depends. Entire geometries that expand far beyond the tile boundaries (not sure how far; values < 0 or >= 4096 are "invisible" but considered when rendering; this may be how feature rendering is isolated to the visible contents of a tile) are valid (if inefficient).

For purely analytical purposes, the 2 obvious options seem like a) full geometries or b) geometries clipped to tile bounds.

For cartographic purposes (typically labels / markers), Mapnik (the original reference implementation) supports per-layer buffers (measured in "pixels") that are used to expand a tile's boundaries and clip features (no reason other than convenience that they can't be per-feature buffers). When layers are buffered, features will be duplicated between tiles and considered when doing collision / duplicate label detection during the render step (I'm mainly familiar with raster rendering). For labeled points that are "off-screen" (relative to a tile being rendered), the label can be rasterized partially off-canvas and it will line up with a neighboring label that has been cropped.

(If my assumption about "invisible coordinates" holds, it may be possible "clip" geometries during the command generation process by dropping coordinates out of the stream when they're outside the buffered tile boundary (in screen coordinates), except when crossing one of the corners.)

Reply:
I'd forgotten about the 3x3 clipping done by Mapbox + Mapzen, referenced in https://github.com/geotrellis/vectorpipe/issues/11.